### DISCUSSION TITLE
Cross Referencing Issue

### DISCUSSION THREAD

**Discussion by Danfs64** [eclipse-langium/langium#798] [Q&A] [answered]
This is a minimal sample demonstrating a problem I'm having:

```langium
entry App: (elem+=Element)*;

Element: (A|B|C);

A: "A" name=ID;

B: "B" name=ID 'with' d=D;

C: "C" name=ID 'ref' ref=[Ref];

D: name=ID;

Ref: (A|D);
```

With the grammar declared above, the following code:
```
A ei
B bee with dee
C cee ref dee
```

Gives an error saying it `could not resolve the reference to Ref named 'dee'`
If instead of `dee` I use `ei`, it works properly, referencing the `A` node

I don't know if this is a bug or something I missed while developing my grammar, so instead of opening an Issue, I'm asking here first if anyone knows why that is happening

**Edit: Also, on the langium code example, it also gives a `This rule is declared but never referenced` on Ref

**Comment by luan-xiaokun** [comment, marked as answer]
I've had this problem before, your grammar is fine, the solution is to implement your own ScopeComputation (a ScopeProvider can also do this I guess). I think the [document](https://langium.org/docs/document-lifecycle/) explains how langium resolves cross reference very well, you should probably check it out first.

According to my understanding, the problem here is that `dee` is not visible to the node `C` (it's not in the scope of `C`), because `dee` is in the subtree of `B`, which is a sibling of `C`. You need to put it in the scope of `C`, so the linker can resolve the reference. [This](https://github.com/langium/langium-website/pull/79) includes a guide for scoping in langium, which is a good starting point.

> Also, on the langium code example, it also gives a `This rule is declared but never referenced` on Ref

~~I think using two different properties for `A` and `D` in this case would be helpful.~~ (See Mark's comment below)

**Reply by msujew** [reply]
Correct on the scoping parts, thanks for the explanation. Note that references such as `refA=[A] |  refD=[D]` won't resolve, since they lead to a parser ambiguity instead. Changing the `Ref` rule to a type such as `type Ref = A | D;` does the trick though.

**Reply by luan-xiaokun** [reply]
Yeah, I was wrong about the `Ref` rule. I will strike it out.

**Reply by Danfs64** [reply]
Oh well, the Document Lifecycle did mention that children of a non-root node aren't visible outside of that node sub-tree, must've missed it when I read it weeks ago.

That said, after some digging, I did managed to figure out how to extend the ScopeComputation to do what I needed (which in my case was: putting the `D` nodes on the global scope). Thanks for the help both of you 😄

**Reply by joswarmer** [reply]
I just read the answer with the type `type Ref = A | D;`.  This was helpful for me to solve a problem I had.
Then I started looking for more explanation of the type in the manual but could not find anything. However, I did find an example where an `interface` is defined. 
I get the strong feeling that typescript types and interfaces are allowed inside langium grammar files, but I cannot find this anywhere.
Am I missing something, or is this something that is missing from the documentation?

**Reply by msujew** [reply]
We have some explanation about [declared types here](https://langium.org/docs/reference/semantic-model/#declared-types).

**Reply by joswarmer** [reply]
Thanks, that is very helpful documentation.
